src/libstd/fs.rs

```diff
@@ -0,0 +1,1474 @@
+    /// * Attempting to open a file with access that the user lacks
+    ///   permissions for
+    /// * Filesystem-level errors (full disk, etc)
+    pub fn open_opts<P: AsPath + ?Sized>(path: &P, opts: &OpenOptions)
+                                         -> io::Result<File> {
+        let path = path.as_path();
+        let inner = try!(fs_imp::File::open(path, &opts.0));
+
+        // On *BSD systems, we can open a directory as a file and read from
+        // it: fd=open("/tmp", O_RDONLY); read(fd, buf, N); due to an old
+        // tradition before the introduction of opendir(3).  We explicitly
+        // reject it because there are few use cases.
+        if cfg!(not(any(target_os = "linux", target_os = "android"))) &&
+           try!(inner.file_attr()).is_dir() {
+            Err(Error::new(ErrorKind::InvalidInput, "is a directory", None))
+        } else {
+            Ok(File { path: path.to_path_buf(), inner: inner })
+        }
+    }
+
+    /// Attempts to open a file in read-only mode.
+    ///
+    /// See the `File::open_opts` function for more details.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if `path` does not already exist.
+    /// Other errors may also be returned according to `File::open_opts`.
+    pub fn open<P: AsPath + ?Sized>(path: &P) -> io::Result<File> {
+        File::open_opts(path, OpenOptions::new().read(true))
+    }
+
+    /// Creates a open a file in write-only mode.
+    ///
+    /// This method will attempt to open a new file, truncating it if it already
+    /// exists.
+    ///
+    /// See the `File::open_opts` function for more details.
+    pub fn create<P: AsPath + ?Sized>(path: &P) -> io::Result<File> {
+        File::open_opts(path, OpenOptions::new().write(true).create(true)
+                                                .truncate(true))
+    }
+
+    /// Returns the original path that was used to open this file.
+    pub fn path(&self) -> &Path {
+        &*self.path
+    }
+
+    /// Attempt to sync all OS-internal metadata to disk.
+    ///
+    /// This function will attempt to ensure that all in-core data reaches the
+    /// filesystem before returning.
+    pub fn sync_all(&mut self) -> io::Result<()> {
+        self.inner.fsync()
+    }
+
+    /// This function is similar to `sync_all`, except that it may not
+    /// synchronize file metadata to the filesystem.
+    ///
+    /// This is intended for use cases that must synchronize content, but don't
+    /// need the metadata on disk. The goal of this method is to reduce disk
+    /// operations.
+    ///
+    /// Note that some platforms may simply implement this in terms of
+    /// `sync_all`.
+    pub fn sync_data(&mut self) -> io::Result<()> {
+        self.inner.datasync()
+    }
+
+    /// Either truncates or extends the underlying file, updating the size of
+    /// this file to become `size`. This is equivalent to unix's `truncate`
+    /// function.
```

Comment:
This note of equivalency has no place in the supposed to be short one sentence description.